### PR TITLE
CI: make sure Paparazzi test failures stop the test flow

### PR DESCRIPTION
**Pull request by jmartinesp**
Changes to CI:
- Run Paparazzi tests just once when `koverMergedReport` is used.
- Make sure a Paparazzi failure breaks the test flow and blocks merging.
- Change `./gradlew check` call in `quality.yml` to custom `runQualityChecks` task.

This should also improve overall build times in the test and quality flows and avoid duplicate test runs.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -21,6 +22,27 @@ jobs:
       cancel-in-progress: true
     steps:
       - uses: actions/checkout@v3
+        with:
+          lfs: 'true'
```
Comment (bmarty) [inline .github/workflows/tests.yml:26]: `lfs: 'true'` was maybe the only thing which was missing to make the screenshot test not failing, WDYT?

File: tests/uitests/src/test/kotlin/io/element/android/tests/uitests/ScreenshotTest.kt [hunk 2]
```diff
@@ -67,7 +67,7 @@ class ScreenshotTest {
 
     @get:Rule
     val paparazzi = Paparazzi(
-        maxPercentDifference = 0.0,
+        maxPercentDifference = 0.1,
```
Comment (bmarty) [inline tests/uitests/src/test/kotlin/io/element/android/tests/uitests/ScreenshotTest.kt:70]: https://github.com/vector-im/element-android/commit/28f4e79b43fa10bcd228c1da2bf19bcd290b4e46

File: .github/workflows/tests.yml (hunk 1 above)
Comment (bmarty) [reply .github/workflows/tests.yml:26]: OK, so this is not enough, we have to explicitly invoke `verifyPaparazziDebug` task for the test to fail :/

File: tests/uitests/build.gradle.kts [hunk 3]
```diff
@@ -42,3 +42,7 @@ dependencies {
     allLibraries()
     allFeatures()
 }
+
+tasks.withType<Test> {
+    enabled = project.hasProperty("runPaparazzi")
+}
```
Comment (bmarty) [inline tests/uitests/build.gradle.kts:48]: It seems that [the coverage is not computed](https://github.com/vector-im/element-x-android/actions/runs/4290584239/jobs/7474748317) if we exclude this module from the global `test` task. So I would remove this `runPaparazzi` trick.

**Review by bmarty — COMMENTED** [review]
Thanks, please see my comment about `runPaparazzi`.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (jmartinesp) [reply .github/workflows/tests.yml:26]: Yep. Also, in my case the tolerance had to be > `0.0` to work, as there are very tiny difference in maybe 6-8 pixels between my generated screenshots and the ones from the CI. It might be due to some color callibration of the OS? I'm not sure, to be honest, but they didn't match 100%.

Maybe something like `0.01` could work?

File: tests/uitests/build.gradle.kts (hunk 3 above)
Comment (jmartinesp) [reply tests/uitests/build.gradle.kts:48]: Then we'll have to run it twice to make it fail, or maybe use some other trick to detect a failure in verifyPaparazzi like verifying if the `failures` folder exists?

Comment (bmarty) [reply tests/uitests/build.gradle.kts:48]: Yes, it's a shame... let's run them twice for now...

Comment (bmarty) [reply tests/uitests/build.gradle.kts:48]: Paparazzi not verifying by default seems to be by design: https://github.com/cashapp/paparazzi/issues/701#issuecomment-1404295807 . The workaround to change this behavior could not work IIUC.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -21,8 +22,27 @@ jobs:
...
+        with:
+          name: reports
+          path: tests/uitests/build/reports/tests/testDebugUnitTest/
+          retention-days: 1
+
+      - name: Upload Screenshot Failure Differences
+        uses: actions/upload-artifact@v3
+        if: failure()
+        with:
+          name: reports
```
Comment (bmarty) [inline .github/workflows/tests.yml:43]: maybe use a distinct name?

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -21,8 +22,27 @@ jobs:
...
+          path: tests/uitests/out/failures/
+          retention-days: 1
```
Comment (bmarty) [inline .github/workflows/tests.yml:45]: Maybe keep the artifact a bit longer?

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -20,19 +21,50 @@ jobs:
...
+      - name: ⏬ Checkout with LFS
+        uses: actions/checkout@v3
+        with:
+          lfs: 'true'
+
+      - name: ⚙️ Run unit & screenshot tests, generate kover report
+        run: ./gradlew koverMergedReport $CI_GRADLE_ARG_PROPERTIES -Pci-build=true
+
+      - name: 📈 Verify coverage
+        run: ./gradlew koverMergedVerify $CI_GRADLE_ARG_PROPERTIES -Pci-build=true
```
Comment (bmarty) [inline .github/workflows/tests.yml:33]: `koverMergedReport` and `koverMergedVerify` will both run the tests IIRC. It's a shame there is not a single task to do both without running the test twice...

Comment (jmartinesp) [reply .github/workflows/tests.yml:33]: But we just launched `verifyPaparazzi` from the previous `koverMergedReport`, so if I'm not mistaken it should be up to date and not run again. At least, I don't see any delay when running these, I'll double check just in case.

**Review by bmarty — APPROVED** [review]
Thanks!
I see that you have recorded the screenshots?

**Comment by jmartinesp** [comment]
> Thanks! I see that you have recorded the screenshots?

Yes, I recorded them again to make sure the `maxPercentDifference = 0.01` was enough to make them work even if the OS used to record the images doesn't match the one that verifies them.

**Comment by jmartinesp** [comment]
I just force pushed to try to force codecov to run. For some reason, it's not running in any of my PRs since a couple of days ago...
